poc_mobile_verifier/src/owner_shares.rs

```diff
@@ -1,94 +1,170 @@
-            *hotspot_shares
-                .entry(heartbeat.hotspot_key.clone())
-                .or_default() += heartbeat.reward_weight;
-        }
+impl RadioShares {
+    pub fn push(&mut self, share: RadioShare) {
+        self.shares.push(share);
+    }
 
-        let filtered_shares = hotspot_shares
-            .into_iter()
-            .map(|(pubkey, mut shares)| {
-                let speedmultiplier = speedtests
-                    .get_average(&pubkey)
-                    .map_or(Decimal::ZERO, |avg| avg.reward_multiplier());
-                shares *= speedmultiplier;
-                (pubkey, shares)
-            })
-            .filter(|(_pubkey, shares)| shares > &Decimal::ZERO)
-            .collect();
+    pub fn total_shares(&self) -> Decimal {
+        self.shares
+            .iter()
+            .fold(Decimal::ZERO, |sum, radio_share| sum + radio_share.amount)
+    }
 
-        let (owner_shares, _missing_owner_shares) =
-            follower_service.owner_shares(filtered_shares).await?;
+    pub fn into_iter(self) -> impl Iterator<Item = RadioShare> {
+        self.shares.into_iter()
+    }
+}
 
-        let owner_emissions =
-            OwnerEmissions::new(owner_shares, epoch.start, epoch.end - epoch.start);
+#[derive(Default)]
+pub struct OwnerShares {
+    pub shares: HashMap<PublicKey, RadioShares>,
+}
 
-        let mut rewards = owner_emissions
-            .into_inner()
-            .into_iter()
-            .map(|(owner, amt)| proto::SubnetworkReward {
-                account: owner.to_vec(),
-                amount: u64::from(amt),
+const REWARDS_PER_SHARE_PREC: u32 = 9;
+const MOBILE_SCALE: u32 = 100_000_000;
+
+impl OwnerShares {
+    pub fn total_shares(&self) -> Decimal {
+        self.shares
+            .iter()
+            .fold(Decimal::ZERO, |sum, (_, radio_shares)| {
+                sum + radio_shares.total_shares()
             })
-            .collect::<Vec<_>>();
-
-        rewards.sort_by(|a, b| {
-            a.account
-                .cmp(&b.account)
-                .then_with(|| a.amount.cmp(&b.amount))
-        });
-
-        Ok(Self {
-            epoch: epoch.clone(),
-            rewards,
-        })
     }
 
-    pub async fn write(
+    pub fn into_radio_shares(
         self,
-        subnet_rewards_tx: &file_sink::MessageSender,
-    ) -> file_store::Result<oneshot::Receiver<file_store::Result>> {
-        file_sink_write!(
-            "subnet_rewards",
-            subnet_rewards_tx,
-            proto::SubnetworkRewards {
-                start_epoch: self.epoch.start.timestamp() as u64,
-                end_epoch: self.epoch.end.timestamp() as u64,
-                rewards: self.rewards,
-            }
+        epoch: &'_ Range<DateTime<Utc>>,
+    ) -> impl Iterator<Item = proto::RadioRewardShare> + '_ {
+        let total_shares = self.total_shares();
+        let total_rewards = get_scheduled_tokens(epoch.start, epoch.end - epoch.start).unwrap();
```

Comment:
seems like we should do something more than just `unwap` here

Reply:
Maybe an `expect`? This error can only occur if the system clock is broken

Reply:
`expect` feels better, at least we make sure the string clearly tells us what happened.

Reply:
why have either? It _crashes_ the current thread, with unknown side effects.. Why not return a Result of an Impl?

Reply:
yeah might as well make it an error